File: src/compiler/scala/tools/util/AbstractTimer.java
/*     ____ ____  ____ ____  ______                                     *\
**    / __// __ \/ __// __ \/ ____/    SOcos COmpiles Scala             **
**  __\_ \/ /_/ / /__/ /_/ /\_ \       (c) 2002, LAMP/EPFL              **
** /_____/\____/\___/\____/____/                                        **
\*                                                                      */

// $Id$

package scala.tools.util;

import java.util.ArrayList;

/**
 * This abstract class implements the collection of timings. How the
 * collected timings are issued has to be implemented in subclasses.
 */
public abstract class AbstractTimer implements Timer {

    //########################################################################
    // Private Fields

    /** A stack for maintaining start times */
    private final ArrayList starts = new ArrayList();

    //########################################################################
    // Public Methods

    /** Issues a timing information (duration in milliseconds). */
    public abstract void issue(String message, long duration);

    /** Starts a new timer. */
    public void start() {
        starts.add(new Long(System.currentTimeMillis()));
    }

    /** Ends the current timer. */
    public void stop(String message) {
        long stop = System.currentTimeMillis();
        long start = ((Long)starts.remove(starts.size() - 1)).longValue();
        issue(message, stop - start);
    }

    /** Drops the current timer. */
    public void drop() {
        starts.remove(starts.size() - 1);
    }

    //########################################################################
}

File: src/compiler/scala/tools/util/AbstractTimer.scala
/*                     __                                               *\
**     ________ ___   / /  ___     Scala API                            **
**    / __/ __// _ | / /  / _ |    (c) 2002-2011, LAMP/EPFL             **
**  __\ \/ /__/ __ |/ /__/ __ |    http://scala-lang.org/               **
** /____/\___/_/ |_/____/_/ | |                                         **
**                          |/                                          **
\*                                                                      */


package scala.tools.util

import compat.Platform.currentTime
import scala.collection.mutable.Stack

/**
 * This abstract class implements the collection of timings. How the
 * collected timings are issued has to be implemented in subclasses.
 *
 * @author Philippe Altherr
 * @version 1.0
 */
abstract class AbstractTimer {

  //########################################################################
  // Private Fields

  /** A stack for maintaining start times */
  private val starts = new Stack[Long]()

  //########################################################################
  // Public Methods

  /** Issues a timing information (duration in milliseconds). */
  def issue(message: String, duration: Long): Unit

  /** Starts a new timer. */
  def start() {
    starts push currentTime
  }

  /** Ends the current timer. */
  def stop(message: String) {
    val stop = currentTime
    issue(message, stop - starts.pop)
  }

  /** Drops the current timer. */
  def drop() {
    starts.pop
  }

    //########################################################################
}
